src/box/tx_memory.c

```diff
@@ -0,0 +1,195 @@
+#include <lib/core/say.h>
+#include "tx_memory.h"
+#include "small/mempool.h"
+#include "small/region.h"
+#include "histogram.h"
+#include "txn.h"
+
+const char *TXN_ALLOC_TYPE_STRS[] = {
+	"STATEMENTS",
+	"SAVEPOINTS",
+	"USER DATA",
+	"REDO LOGS",
+	"TRIGGERS",
+	"TIMERS",
+	"JOURNAL ENTRIES"
+};
+
+static_assert(lengthof(TXN_ALLOC_TYPE_STRS) == TXN_ALLOC_MAX,
+	      "TXN_ALLOC_TYPE_STRS does not match TXN_ALLOC_TYPE");
+
+static void
+tx_track_allocation(struct tx_memory_manager *stat, struct txn *txn,
+		    uint64_t alloc_size, size_t alloc_type, bool deallocate)
+{
+	assert(alloc_type < stat->alloc_max);
+	assert(txn != NULL);
+
+	histogram_discard(stat->stats_storage[alloc_type].hist,
+			  txn->mem_used->total[alloc_type]);
+	if (!deallocate) {
+		txn->mem_used->total[alloc_type] += alloc_size;
+		stat->stats_storage[alloc_type].total += alloc_size;
+	} else {
+		assert(txn->mem_used->total[alloc_type] >= alloc_size);
+		assert(stat->stats_storage[alloc_type].total >= alloc_size);
+		txn->mem_used->total[alloc_type] -= alloc_size;
+		stat->stats_storage[alloc_type].total -= alloc_size;
+	}
+	histogram_collect(stat->stats_storage[alloc_type].hist,
+			  txn->mem_used->total[alloc_type]);
+}
+
+void
+tx_memory_register_txn(struct tx_memory_manager *stat, struct txn *txn)
+{
+	assert(txn != NULL);
+	assert(txn->mem_used == NULL);
+	assert(txn->given_region_used == 0);
+
+	struct txn_mem_used *mem_used =
+		region_aligned_alloc(&txn->region,
+				     sizeof(struct txn_mem_used)  +
+					     sizeof(uint64_t) * stat->alloc_max,
+				     alignof(*(txn->mem_used)));
+	memset(mem_used, 0, sizeof(struct txn_mem_used)  +
+		sizeof(uint64_t) * stat->alloc_max);
+	txn->mem_used = mem_used;
+	for (size_t i = 0; i < stat->alloc_max; ++i) {
+		assert(mem_used->total[i] == 0);
+		histogram_collect(stat->stats_storage[i].hist,0);
+	}
+	stat->txn_num++;
+}
+
+void
+tx_memory_clear_txn(struct tx_memory_manager *stat, struct txn *txn)
+{
+	assert(txn != NULL);
+	assert(txn->mem_used != NULL);
+	assert(txn->given_region_used == 0);
+
+	// Check if txn does not owe any mempool allocation.
+	// In this case all the tracked allocation are from region
+	// and we will delete them via region_truncate.
+	assert(txn->mem_used->mempool_total == 0);
+	for (size_t i = 0; i < stat->alloc_max; ++i) {
+		tx_track_allocation(stat, txn, txn->mem_used->total[i], i, true);
+	}
+	txn->mem_used = NULL;
+	region_truncate(&txn->region, sizeof(struct txn));
+	assert(stat->txn_num > 0);
+	stat->txn_num--;
+}
+
+void *
+tx_memory_mempool_alloc(struct tx_memory_manager *stat, struct txn *txn,
+			struct mempool* pool, size_t alloc_type)
+{
+	assert(pool != NULL);
+	assert(txn != NULL);
+	assert(alloc_type < stat->alloc_max);
+
+	struct mempool_stats pool_stats;
+	mempool_stats(pool, &pool_stats);
+	tx_track_allocation(stat, txn, pool_stats.objsize, alloc_type, false);
+#ifndef NDEBUG
+	txn->mem_used->mempool_total += pool_stats.objsize;
+#endif
+	return mempool_alloc(pool);
```

Comment:
In case mempool_alloc fails you can get leak according to your tracker. So I guess we should firstly check that mempool was successful and only then account memory usage.

Reply:
FIxed